### PR TITLE
Match updated wallet spec

### LINKED ISSUE #32: Misc key derivation changes
From last week's email chain, we ought:

* Remove the optional 'password' salt throughout
* Remove the 'revision' salt throughout
* Not use the ship number to salt in childSeedFromSeed
* Fix all seeds at 256 bits
* Use an explicit out: 32 bytes argument to argon2
* Use an explicit derivation path of `m/44'/60'/0'/0/0`

See the updated wallet spec for details.

### PR DESCRIPTION
**Pull request by jtobin**
This is a bigger PR than I'd normally want to make, but there were a lot of changes to make, and I want to get them all in before this week is out.  It thus wound up being easier to just go into bull-in-china-shop mode than be particularly surgical here.  So, this covers a bit of ground.  From the CHANGELOG:

* Fix all seeds at 256 bits.
* Always derive BIP32 nodes according to the path m/44'/60'/0'/0/0.
* Remove the optional 'password' salt.
* Render and handle (non-network) seeds as BIP39 mnemonics.
* Displays corresponding Ethereum addresses in derived BIP32 nodes.
* Replaces XOR-based 2-3 sharding with concat-based 2-3 sharding.  Shards
   are also now reported in the 'shards' property of the final wallet.
* Add an 'ownership' seed, plus slight changes to other seed type names.
* Miscellaneous housekeeping and simplification throughout.

I believe this resolves #32, #27, #26, #25, and #22, and takes us to the [updated wallet spec](https://docs.google.com/document/d/1F_fx6P0riv5LjYI3DyLfL7L7UDkjm1n5d3nJWrXln94/edit) and also addresses various concerns from the auditors' reports.

The housekeeping/simplification axed a few hundred lines of code in `src/index.js`, and the test suite slimmed down as a result.  I also made some improvements there -- added some property tests, improved whole-wallet unit tests, etc.

For reviewing: I would ignore the diff and just read/skim the new `src/index.js` from top to bottom, perhaps with reference to the updated spec (which it maps to very clearly).  It's not very long, and should be pretty clear throughout.

### CODE REVIEW

**Comment by gavinpatkinson** [comment]
There are issues with reserved word usage here – `public` and `private`, I will push fixes.

**Comment by gavinpatkinson** [comment]
This is actually beautiful work

File: src/index.js [hunk 1]
```diff
@@ -1,567 +1,363 @@
...
-    ? `${type}-${revision}-${ship}`
-    : `${type}-${revision}`;
-
-  const childSeed = await hash(seed, salt, defaultTo(password, ''));
-
-  return childSeed.slice(0, seed.byteLength || seed.length);
-};
+  const { seed, type, ship, revision } = config
+  const salt = lodash.isNull(ship) ? '' : `${ship}`
+  const hash = await sha256(seed, type, salt, `${revision}`)
```
Comment (Fang-) [inline src/index.js]: This isn't doing dashes in between the salt parameters. It won't distinguish between ship `5` with revision `10`, and ship `51` with revision `0`.

Pretty sure we want to do `type-revision-ship` like we used to, and then use `0` as the defaults for anything not supplied. Seems correct, @msutherl ?

File: src/index.js [hunk 2]
```diff
@@ -1,567 +1,363 @@
...
+  return type !== CHILD_SEED_TYPES.NETWORK
+    ? bip39.entropyToMnemonic(hash)
+    : Buffer.from(hash).toString('hex')
+}
```
Comment (Fang-) [inline src/index.js]: Maybe we want to pull this out of the derivation function? But maybe not...

File: src/index.js [hunk 3]
```diff
@@ -1,567 +1,363 @@
...
 }
 
 
 
 /**
- * Combine pieces of a sharded buffer together to recover the original buffer.
- * @param  {Array of Array of integers}  shards a collection of shards
- * @return {Buffer} the unsharded buffer
+ * Convert a hex-encoded Ed25519-variant Urbit public network key into an
+ * Ethereum address.
```
Comment (Fang-) [inline src/index.js]: Lol why. This is neat, but doesn't actually need to happen. Urbit keys are for Urbit-use only, representing seed, pub and prv as hex strings is totally fine here.

File: src/index.js [hunk 4]
```diff
@@ -1,567 +1,363 @@
...
+
+  const crypt_pub = buf2hex(crypt.publicKey.reverse())
+  const auth_pub = buf2hex(auth.publicKey.reverse())
 
   return {
     crypt: {
       private: buf2hex(c.reverse()),
-      public: buf2hex(crypt.publicKey.reverse()),
+      public: crypt_pub,
+      address: addressFromNetworkPublic(crypt_pub)
```
Comment (Fang-) [inline src/index.js]: As per another comment below: no point in deriving Ethereum addresses from Urbit keys.

File: src/index.js [hunk 5]
```diff
@@ -1,567 +1,363 @@
...
-  if (shardsConsistent(combinePatq, patq, ob.eqPatq, shards)) {
-    return shards;
-  } else {
-    // shards must always be consistent, so this should be unreachable
-    /* istanbul ignore next */
-    throw("shardPatq: inconsistent shards -- please report this as a bug");
-  }
+const addressFromSecp256k1Public = pub => {
+  const hashed = util.keccak256(Buffer.from(pub, 'hex'))
+  const addr = util.addHexPrefix(hashed.slice(12).toString('hex'))
```
Comment (Fang-) [inline src/index.js]: The `.slice(12)` reads a bit weird here, might want to put a comment to clarify that this is taking the last 20 bytes.

File: src/index.js [hunk 6]
```diff
@@ -1,567 +1,363 @@
...
  */
-const shardWallet = wallet => {
-  const walletCopy = lodash.cloneDeep(wallet);
-  const sharded = shardPatq(walletCopy.ticket)
-  walletCopy.ticket = sharded;
-  return walletCopy;
-}
+const generateWallet = async config => {
+  const { ticket } = config
+  const ship = 'ship' in config ? config.ship : null
```
Comment (Fang-) [inline src/index.js]: As with the salt, I think we default to `0` rather than complete absence. @msutherl ?

**Comment by Fang-** [comment]
(I'm not joking about the semicolons btw, I would like to see those back in. Of course, I am a mere mortal, making demands in front of a God...)

File: src/index.js (hunk 6 above)
Comment (gavinpatkinson) [reply src/index.js]: Yes, this was the conclusion from the other day.

File: src/index.js (hunk 1 above)
Comment (jtobin) [reply src/index.js]: Yeah, agree that's better.  But we should specify the salt format concretely in the spec, which at present reads along the lines of what I've done here.  /willfix in the meantime.

File: src/index.js (hunk 3 above)
Comment (jtobin) [reply src/index.js]: Yeah, I thought.. do we actually need this?  And then erred on the side of caution.. 😁   Will strip out.

**Comment by jtobin** [comment]
@Fang- Will make all changes you requested.  Except, of course, for the semicolons. 😁

(If I were to adopt a particular style guide it would probably be [this one](https://standardjs.com/), just because it appears to be.. uh, standard (boy they really nailed it with the name choice), and which just happens to require no semicolons.)

**Comment by Fang-** [comment]
> Except, of course, for the semicolons. 😁

https://github.com/urbit/keygen-js/pull/2#issuecomment-416728380

**Comment by jtobin** [comment]
Thanks for the reviews gents!

**Comment by msutherl** [comment]
I put in an extra review for good measure – beautiful work

**Pull request merged**